hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/stack/SavepointStackImpl.java

```diff
@@ -531,20 +535,39 @@ public HandleOutput buildHandleOutput(
                 indexOfTopLevelRecord = i;
                 topLevelNonce = builder.transactionID().nonce();
                 idBuilder = builder.transactionID().copyBuilder();
+                isBatch = builder.functionality() == ATOMIC_BATCH;
                 break;
             }
         }
         int nextNonceOffset = 1;
+        var parentConsensusTime = consensusTime;
         for (int i = 0; i < n; i++) {
             final var builder = builders.get(i);
             final var nonceOffset =
                     switch (builder.category()) {
-                        case USER, SCHEDULED, NODE -> 0;
+                        case USER, SCHEDULED, NODE, BATCH -> 0;
                         case PRECEDING, CHILD -> nextNonceOffset++;
                     };
             final var txnId = builder.transactionID();
             // If the builder does not already have a transaction id, then complete with the next nonce offset
             if (txnId == null || TransactionID.DEFAULT.equals(txnId)) {
+                if (i > indexOfTopLevelRecord && isBatch) {
+                    if (builder.category() == PRECEDING) {
+                        for (int j = i + 1; j < n; j++) {
+                            if (builders.get(j).category() == BATCH) {
+                                idBuilder = builders.get(j).transactionID().copyBuilder();
+                                break;
+                            }
+                        }
+                    } else if (builder.category() == CHILD) {
+                        for (int j = i - 1; j >= indexOfTopLevelRecord; j--) {
```

Comment:
For clarity of intent, `j > indexOfTopLevelRecord` (top-level dispatch category cannot be `BATCH`).

Reply:
Fixed.